File: architecture-skills/design-patterns/examples/memento/memento.cs
// Memento pattern in C#.
//
// A record snapshot keeps the memento immutable. Nesting it inside the
// editor with a private constructor means only the originator can create
// or read one; the history just stacks opaque values.

using System;
using System.Collections.Generic;

public sealed class TextEditor
{
    private string _text = "";
    private int _cursor;

    // Memento: creatable and readable only by TextEditor.
    public sealed record Snapshot
    {
        internal string Text { get; }
        internal int Cursor { get; }

        internal Snapshot(string text, int cursor) => (Text, Cursor) = (text, cursor);
    }

    public void Type(string insert)
    {
        _text = _text.Insert(_cursor, insert);
        _cursor += insert.Length;
    }

    public Snapshot Save() => new(_text, _cursor);

    public void Restore(Snapshot s) => (_text, _cursor) = (s.Text, s.Cursor);

    public override string ToString() => $"'{_text}' (cursor at {_cursor})";
}

public sealed class History
{
    private readonly Stack<TextEditor.Snapshot> _snapshots = new();

    public void Push(TextEditor.Snapshot s) => _snapshots.Push(s);

    public bool TryPop(out TextEditor.Snapshot? s) => _snapshots.TryPop(out s);
}

public static class Program
{
    public static void Main()
    {
        var editor = new TextEditor();
        var history = new History();

        editor.Type("Hello");
        history.Push(editor.Save());
        editor.Type(", world");
        history.Push(editor.Save());
        editor.Type("!!!");
        Console.WriteLine($"current: {editor}");

        for (int i = 0; i < 2 && history.TryPop(out var s); i++)
        {
            editor.Restore(s!);
            Console.WriteLine($"undo:    {editor}");
        }
    }
}

File: architecture-skills/design-patterns/examples/memento/memento.php
<?php
// Memento pattern in PHP.
//
// Readonly properties make the snapshot immutable once created; the
// history stacks snapshots without interpreting them. For crude
// whole-object snapshots, clone (with __clone for deep copies) is a
// lighter alternative.

declare(strict_types=1);

/** Memento: immutable snapshot of editor state. */
final class Snapshot
{
    public function __construct(
        public readonly string $text,
        public readonly int $cursor,
    ) {
    }
}

/** Originator. */
final class TextEditor
{
    private string $text = '';
    private int $cursor = 0;

    public function type(string $insert): void
    {
        $this->text = substr($this->text, 0, $this->cursor)
            . $insert
            . substr($this->text, $this->cursor);
        $this->cursor += strlen($insert);
    }

    public function save(): Snapshot
    {
        return new Snapshot($this->text, $this->cursor);
    }

    public function restore(Snapshot $snapshot): void
    {
        $this->text = $snapshot->text;
        $this->cursor = $snapshot->cursor;
    }

    public function __toString(): string
    {
        return "'{$this->text}' (cursor at {$this->cursor})";
    }
}

/** Caretaker: stores snapshots but never inspects them. */
final class History
{
    /** @var list<Snapshot> */
    private array $snapshots = [];

    public function push(Snapshot $snapshot): void
    {
        $this->snapshots[] = $snapshot;
    }

    public function pop(): ?Snapshot
    {
        return array_pop($this->snapshots);
    }
}

$editor = new TextEditor();
$history = new History();

$editor->type('Hello');
$history->push($editor->save());
$editor->type(', world');
$history->push($editor->save());
$editor->type('!!!');
echo "current: $editor\n";

for ($i = 0; $i < 2; $i++) {
    $snapshot = $history->pop();
    if ($snapshot !== null) {
        $editor->restore($snapshot);
        echo "undo:    $editor\n";
    }
}
